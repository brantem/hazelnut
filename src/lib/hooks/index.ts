export * from 'lib/hooks/useSearch';
export * from 'lib/hooks/useModal';
