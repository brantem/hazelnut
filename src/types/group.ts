export type Group = {
  id: string;
  title: string;
  color: string;
  minimized: boolean;
  createdAt: number;
};
