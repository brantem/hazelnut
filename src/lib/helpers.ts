import days from 'data/days';
import { Day } from 'types/shared';

export const getMinutesFromTime = (time: string) => {
  const [hour, minute] = time.split(':');
  return parseInt(hour) * 60 + parseInt(minute);
};

export const getCurrentDay = (): Day => [...days.slice(6), ...days.slice(0, -1)][new Date().getDay()];
