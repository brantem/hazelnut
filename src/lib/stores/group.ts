import create from 'zustand';

import { Group } from 'types/group';

type GroupState = {
  group: Group | null;
  clear: () => void;

  isSaveOpen: boolean;
  showSave: (group?: Group | null) => void;

  isSettingsOpen: boolean;
  showSettings: (group: Group) => void;

  hide: () => void;
};

export const useGroupStore = create<GroupState>()((set) => ({
  group: null,
  clear: () => set({ group: null }),

  isSaveOpen: false,
  showSave: (group = null) => set((state) => ({ group: group || state.group, isSaveOpen: true })),

  isSettingsOpen: false,
  showSettings: (group) => set({ group, isSettingsOpen: true }),

  hide: () => set({ isSaveOpen: false, isSettingsOpen: false }),
}));
