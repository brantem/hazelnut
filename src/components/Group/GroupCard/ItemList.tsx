import { useCallback } from 'react';
import { MinusCircleIcon } from '@heroicons/react/24/outline';
import clsx from 'clsx';

import DeleteButton from 'components/DeleteButton';

import { useItemsStore } from 'lib/stores';
import type { Group } from 'types/group';

type ItemListProps = {
  group: Group;
};

const ItemList = ({ group }: ItemListProps) => {
  const items = useItemsStore(useCallback((state) => state.getItemsByGroupId(group.id), [group.id]));
  const remove = useItemsStore((state) => state.remove);

  return (
    <ol className="space-y-1 pt-2 pb-1" data-testid="group-card-items">
      {items.map((item) => (
        <li data-testid="group-card-items-item" key={item.id} className="flex items-center justify-between space-x-3">
          <span className="truncate">{item.title}</span>

          <DeleteButton
            className={(clicked) => clsx(`rounded-md !p-1 text-sm text-red-500 hover:bg-red-100`, clicked && '!px-2')}
            text={<MinusCircleIcon className="h-5 w-5" />}
            onConfirm={() => remove(item.id)}
          />
        </li>
      ))}
    </ol>
  );
};

export default ItemList;
