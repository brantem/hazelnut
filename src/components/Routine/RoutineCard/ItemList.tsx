import { useState, useCallback } from 'react';
import {
  DndContext,
  closestCenter,
  MouseSensor,
  TouchSensor,
  KeyboardSensor,
  useSensor,
  useSensors,
  UniqueIdentifier,
} from '@dnd-kit/core';
import {
  arrayMove,
  SortableContext,
  useSortable,
  sortableKeyboardCoordinates,
  verticalListSortingStrategy,
} from '@dnd-kit/sortable';
import { restrictToVerticalAxis, restrictToParentElement } from '@dnd-kit/modifiers';
import { CSS } from '@dnd-kit/utilities';
import clsx from 'clsx';

import Checkbox from 'components/Checkbox';

import { useItemsStore, useRoutinesStore, useHistoriesStore } from 'lib/stores';
import { Routine } from 'types/routine';
import { Item as _Item } from 'types/item';

type ItemProps = {
  routine: Routine;
  item: _Item;
  isSortable?: boolean;
};

const Item = ({ routine, item, isSortable }: ItemProps) => {
  const isDone = useHistoriesStore(
    useCallback((state) => state.getIsDone(routine.id, item.id, true), [routine.id, item.id]),
  );
  const save = useHistoriesStore((state) => state.save);
  const { attributes, listeners, setNodeRef, setActivatorNodeRef, transform, transition } = useSortable({
    id: item.id,
  });

  return (
    <li
      ref={setNodeRef}
      className="flex h-8 w-full items-center justify-between space-x-2 pr-1"
      style={{ transform: CSS.Transform.toString(transform), transition }}
      {...attributes}
    >
      {isSortable && (
        <button
          className="-ml-1.5 flex-grow-0 p-1.5"
          ref={setActivatorNodeRef}
          {...listeners}
          data-testid="routine-item-handle"
        >
          <svg viewBox="0 0 20 20" width="12" className="h-4 w-4 text-neutral-500" fill="currentColor">
            <path d="M7 2a2 2 0 1 0 .001 4.001A2 2 0 0 0 7 2zm0 6a2 2 0 1 0 .001 4.001A2 2 0 0 0 7 8zm0 6a2 2 0 1 0 .001 4.001A2 2 0 0 0 7 14zm6-8a2 2 0 1 0-.001-4.001A2 2 0 0 0 13 6zm0 2a2 2 0 1 0 .001 4.001A2 2 0 0 0 13 8zm0 6a2 2 0 1 0 .001 4.001A2 2 0 0 0 13 14z"></path>
          </svg>
        </button>
      )}

      <div className={clsx('w-full', isSortable && 'max-w-[calc(100%-theme(spacing.8))]')}>
        <Checkbox
          label={item.title}
          name={routine.id + '-' + item.id}
          color={routine.color}
          checked={isDone}
          onChange={(e) => save(routine, item, e.target.checked)}
        />
      </div>
    </li>
  );
};

type ItemListProps = {
  routine: Routine;
  isSortable?: boolean;
};

const ItemList = ({ routine, isSortable }: ItemListProps) => {
  const edit = useRoutinesStore((state) => state.edit);
  const items = useItemsStore(useCallback((state) => state.getItemsByIds(routine.itemIds), [routine.itemIds]));
  const sensors = useSensors(
    useSensor(MouseSensor),
    useSensor(TouchSensor),
    useSensor(KeyboardSensor, { coordinateGetter: sortableKeyboardCoordinates }),
  );

  const [itemId, setItemId] = useState<UniqueIdentifier | null>(null);
  const activeIndex = items.findIndex((item) => item.id === itemId);

  if (!items.length) return null;

  return (
    <DndContext
      sensors={sensors}
      collisionDetection={closestCenter}
      onDragStart={({ active }) => active && setItemId(active.id)}
      onDragEnd={({ over }) => {
        setItemId(null);
        if (!over) return;
        const overIndex = items.findIndex((item) => item.id === over.id);
        if (activeIndex === overIndex) return;
        edit(routine.id, { itemIds: arrayMove(items.map((item) => item.id), activeIndex, overIndex) }); // prettier-ignore
      }}
      onDragCancel={() => setItemId(null)}
      modifiers={[restrictToVerticalAxis, restrictToParentElement]}
    >
      <SortableContext items={items} strategy={verticalListSortingStrategy}>
        <ol className="space-y-1 pt-1" data-testid="routine-card-items">
          {items.map((item) => (
            <Item key={item.id} routine={routine} item={item} isSortable={isSortable} />
          ))}
        </ol>
      </SortableContext>
    </DndContext>
  );
};

export default ItemList;
