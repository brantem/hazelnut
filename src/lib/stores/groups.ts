import { useState, useEffect } from 'react';
import create from 'zustand';
import { persist } from 'zustand/middleware';
import { nanoid } from 'nanoid';

import { Group, GroupItem } from 'types/group';

type GroupsState = {
  groups: Group[];
  add: (group: Omit<Group, 'id' | 'items'>) => void;
  addItem: (id: string, item: Omit<GroupItem, 'id'>) => void;
  edit: (id: string, group: Partial<Omit<Group, 'id' | 'items'>>) => void;
  remove: (id: string) => void;
  removeItem: (id: string, itemId: string) => void;
};

const useStore = create<GroupsState>()(
  persist(
    (set) => ({
      groups: [],
      add: (group) => {
        set((state) => ({ groups: [...state.groups, { id: nanoid(), items: [], ...group }] }));
      },
      addItem: (id, item) => {
        set((state) => ({
          groups: state.groups.map((group) =>
            group.id === id ? { ...group, items: [...group.items, { id: nanoid(), ...item }] } : group,
          ),
        }));
      },
      edit: (id, group) => {
        set((state) => ({
          groups: state.groups.map((_group) => (_group.id === id ? { ..._group, ...group } : _group)),
        }));
      },
      remove: (id) => {
        set((state) => ({ groups: state.groups.filter((group) => group.id !== id) }));
      },
      removeItem: (id, itemId) => {
        set((state) => ({
          groups: state.groups.map((group) =>
            group.id === id ? { ...group, items: group.items.filter((item) => item.id !== itemId) } : group,
          ),
        }));
      },
    }),
    { name: 'groups' },
  ),
);

/* c8 ignore start */
const dummy = {
  groups: [],
  add: () => {},
  addItem: () => {},
  edit: () => {},
  remove: () => {},
  removeItem: () => {},
};

// https://github.com/pmndrs/zustand/issues/1145
export const useGroupsStore = ((selector, equals) => {
  const store = useStore(selector, equals);
  const [isHydrated, setHydrated] = useState(false);
  useEffect(() => setHydrated(true), []);
  return isHydrated ? store : selector ? selector(dummy) : dummy;
}) as typeof useStore;
/* c8 ignore stop */
