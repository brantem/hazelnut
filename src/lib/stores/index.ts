import { useEffect } from 'react';

import { itemsStore } from 'lib/stores/items';
import { groupsStore } from 'lib/stores/groups';
import { routinesStore } from 'lib/stores/routines';
import { historiesStore } from 'lib/stores/histories';
import storage from 'lib/stores/storage';

/* c8 ignore start */
export const useLoadStore = () => {
  useEffect(() => {
    (async () => {
      const [items, groups, routines, histories] = await Promise.all([
        storage.getAll('items'),
        storage.getAll('groups'),
        storage.getAll('routines'),
        storage.getAll('histories'),
      ]);

      itemsStore.setState({ items: items.sort((a, b) => a.createdAt - b.createdAt), isReady: true });
      groupsStore.setState({ groups: groups.sort((a, b) => a.createdAt - b.createdAt), isReady: true });
      routinesStore.setState({ routines, isReady: true });
      historiesStore.setState({ histories, selectedDate: localStorage.getItem('history-selected-date') });
    })();
  }, []);
};
/* c8 ignore stop */

export * from 'lib/stores/items';
export * from 'lib/stores/groups';
export * from 'lib/stores/routines';
export * from 'lib/stores/search';
export * from 'lib/stores/histories';
export * from 'lib/stores/modal';
