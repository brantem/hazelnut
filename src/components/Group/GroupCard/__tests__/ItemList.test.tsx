import { render, screen, renderHook, act } from '@testing-library/react';
import '@testing-library/jest-dom';

import ItemList from 'components/Group/GroupCard/ItemList';

import { Group } from 'types/group';
import { useItemsStore, useSearchStore } from 'lib/stores';

const group: Group = {
  id: 'group-1',
  title: 'Group 1',
  color: 'red',
  minimized: false,
};

describe('ItemList', () => {
  beforeAll(() => {
    const items = renderHook(() => useItemsStore());
    act(() => {
      items.result.current.add('group-1', { id: 'item-1', title: 'Item 1' } as Group);
      items.result.current.add('group-2', { id: 'item-2', title: 'Item 2' } as Group);
    });
  });

  it('should render successfully', () => {
    const { container } = render(<ItemList group={group} />);

    expect(container).toMatchSnapshot();
    expect(screen.getByText('Item 1')).toBeInTheDocument();
    expect(screen.queryByText('Item 2')).not.toBeInTheDocument();
  });

  it('should remove item', () => {
    const { result } = renderHook(() => useItemsStore());
    const remove = vi.spyOn(result.current, 'remove');

    render(
      <ItemList
        group={{
          id: 'group-2',
          title: 'Group 2',
          color: 'red',
          minimized: false,
        }}
      />,
    );

    expect(screen.getByText('Item 2')).toBeInTheDocument();
    act(() => screen.getByTestId('delete-button').click());
    act(() => screen.getByTestId('delete-button-confirm').click());
    expect(remove).toHaveBeenCalledWith('item-2');
    expect(screen.queryByText('Item 2')).not.toBeInTheDocument();
  });

  it('should render empty there is no items', () => {
    render(
      <ItemList
        group={{
          id: 'group-3',
          title: 'Group 3',
          color: 'red',
          minimized: false,
        }}
      />,
    );

    expect(screen.queryByTestId('group-card-items')).not.toBeInTheDocument();
  });

  it('should render empty there is no items that contains search value', () => {
    const search = renderHook(() => useSearchStore('items'));

    render(<ItemList group={group} />);

    expect(screen.getByTestId('group-card-items')).toBeInTheDocument();
    act(() => search.result.current.setSearch('a'));
    expect(screen.queryByTestId('group-card-items')).not.toBeInTheDocument();
  });
});
