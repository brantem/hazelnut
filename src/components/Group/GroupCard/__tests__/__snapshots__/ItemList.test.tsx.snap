// Vitest Snapshot v1

exports[`ItemList > should render successfully 1`] = `
<div>
  <ol
    class="space-y-1 pt-1"
    data-testid="group-card-items"
  >
    <li
      class="flex h-8 items-center justify-between space-x-3"
    >
      <span
        class="truncate"
      >
        Item 1
      </span>
      <button
        class="rounded-md p-1 hover:bg-red-100"
        data-testid="group-item-settings"
      >
        <svg
          aria-hidden="true"
          class="h-5 w-5"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 10a1.5 1.5 0 113 0 1.5 1.5 0 01-3 0zM8.5 10a1.5 1.5 0 113 0 1.5 1.5 0 01-3 0zM15.5 8.5a1.5 1.5 0 100 3 1.5 1.5 0 000-3z"
          />
        </svg>
      </button>
    </li>
  </ol>
</div>
`;
